File: 20-valid-parentheses/20-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  const bracketParis = {
    ")": "(",
    "}": "{",
    "]": "["
  };

  const openBrackets = ["(", "{", "["];
  const keepTrackOpen = [];
  
  for (let i = 0, n = s.length; i < n; i++) {
    const char = s[i];
    if (openBrackets.includes(char)) {
      keepTrackOpen.push(char);
    } else if (bracketParis[char] === keepTrackOpen.pop()) {
      continue;
    } else {
      return false;
    }
  }
  
  return (!keepTrackOpen.length);
};

File: 20-valid-parentheses/20-valid-parentheses.rb
def is_valid(s)
  bracket_paris = { ')' => '(', '}' => '{', ']' => '[' }

  open_bracket = ['(', '{', '[']
  keep_track_open = []

  s.each_char do |c|
    if open_bracket.include?(c)
      keep_track_open.push(c)
    elsif bracket_paris[c] == keep_track_open.pop
      next
    else
      return false
    end
  end

  keep_track_open.empty?
end
